readme.md
# ExportCsv

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Total Downloads][ico-downloads]][link-downloads]
[![Build Status][ico-travis]][link-travis]

A Laravel package for hassle free generate CSV files from Database, Export all Data from just one command.

## Installation

Via Composer

``` bash
$ composer require dipenparmar12/exportcsv
```

## Usage

Syntax

> `php artisan csv:export tables=<table-1>,<table-2>`

Example

> `php artisan csv:export tables=users,posts,comments`


Export all tables
> `php artisan csv:export`


Export all tables with `force` option.
> `php artisan csv:export --force`


## Change log

Please see the [changelog](changelog.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Credits

-   [Dipen Parmar](https://dipen.xyz)
-   [All Contributors](../../contributors)

## License

license. Please see the [license file](license.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/dipenparmar12/exportcsv.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/dipenparmar12/exportcsv.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/dipenparmar12/exportcsv/master.svg?style=flat-square
[ico-styleci]: https://styleci.io/repos/12345678/shield

[link-packagist]: https://packagist.org/packages/dipenparmar12/exportcsv
[link-downloads]: https://packagist.org/packages/dipenparmar12/exportcsv
[link-travis]: https://travis-ci.org/dipenparmar12/exportcsv
[link-styleci]: https://styleci.io/repos/12345678
[link-author]: https://github.com/dipenparmar12
[link-contributors]: ../../contributors


changelog.md
# Changelog

All notable changes to `ExportCsv` will be documented in this file.

## Version 1.0

### Added
- Notes...


src/ExportCsv.php
<?php

namespace Dipenparmar12\ExportCsv;

use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\File;
use Illuminate\Support\Facades\Schema;

class ExportCsv
{
    protected $tables = [];

    /**
     * ExportCsv constructor.
     * @param array $tables
     */
    public function __construct(array $tables = [])
    {
        $this->setTables($tables);
    }

    public function export()
    {
        foreach ($this->getTables() as $table) {
            $this->writeCsv($this->getTableData($table), $table);
        }
        return true;
    }

    public function getTables()
    {
        return $this->tables;
    }

    public function setTables($tables = [])
    {
        if ($tables) {
            return $this->tables = $tables;
        }

        $tables = DB::connection()->getDoctrineSchemaManager()->listTableNames();
        foreach ($tables as $table) {
            if (Schema::hasTable($table)) {
                $this->tables[] = $table;
            }
        }
        return $this->tables;
    }

    public function writeCsv($tableData, $filename, $directory = 'database/csv')
    {
        $csv_directory = $this->getDir(base_path() . '/' . $directory);
        $csv_file_path = $csv_directory . '/' . $filename . '.csv';
        $file_handler = fopen($csv_file_path, 'w');
        $headings_row = null;

        if (!$this->tableHasData($tableData)) {
            $this->writeCsvHeading($file_handler, $filename);
            fclose($file_handler);
            return false;
        }

        foreach ($tableData as $tableRow) {
            if (empty($headings_row)) {
                $headings_row = $this->writeCsvRow($file_handler, array_keys($tableRow));
            }
            $this->writeCsvRow($file_handler, $tableRow);
        }

        return fclose($file_handler);
    }

    protected function getDir($path)
    {
        if (!File::exists($path) && File::makeDirectory($path)) {
            return $path;
        }
        return $path;
    }

    protected function tableHasData($table)
    {
        return (is_countable($table) and count($table) > 0);
    }

    protected function writeCsvHeading($file_handler, $table)
    {
        /// Doctrine DBAL Enum issue resolver
        /// Unknown database type enum, MySQL57Platform may not support it #3161
        $conn = DB::connection()->getDoctrineSchemaManager();
        $conn->getDatabasePlatform()->registerDoctrineTypeMapping('enum', 'string');
        /// Ref:1 https://github.com/doctrine/dbal/issues/3161
        /// Ref:2 https://github.com/laravel/framework/issues/1186#issuecomment-17541553

        $column_names = array_keys($conn->listTableColumns($table));
        return fputcsv($file_handler, $column_names);
    }

    protected function writeCsvRow($file, $data)
    {
        return fputcsv($file, $data);
    }

    public function getTableData($table)
    {
        return json_decode(json_encode(DB::table($table)->get()), true);
        //        return DB::table($table)->get()->toArray() ?? [];
    }

    public function tables()
    {
        return $this->tables;
    }
}


src/ExportCsvServiceProvider.php
<?php

namespace Dipenparmar12\ExportCsv;

use Dipenparmar12\ExportCsv\Commands\ExportCsvCommand;
use Illuminate\Support\ServiceProvider;

class ExportCsvServiceProvider extends ServiceProvider
{
    /**
     * Perform post-registration booting of services.
     *
     * @return void
     */
    public function boot()
    {
        // Publishing is only necessary when using the CLI.
        if ($this->app->runningInConsole()) {
            $this->bootForConsole();
        }
    }

    /**
     * Register any package services.
     *
     * @return void
     */
    public function register()
    {
        // Register the service the package provides.
        $this->app->singleton('exportcsv', function ($app) {
            return new ExportCsv;
        });
    }

    /**
     * Get the services provided by the provider.
     *
     * @return array
     */
    public function provides()
    {
        return ['exportcsv'];
    }

    /**
     * Console-specific booting.
     *
     * @return void
     */
    protected function bootForConsole()
    {
        // Registering package commands.
        $this->commands([
            ExportCsvCommand::class
        ]);
    }
}


src/Commands/ExportCsvCommand.php
<?php

namespace Dipenparmar12\ExportCsv\Commands;

use Illuminate\Console\Command;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\File;
use Illuminate\Support\Facades\Schema;

class ExportCsvCommand extends Command
{
    /**
     * The name and signature of the console command.
     *
     * @var string
     */
    protected $signature = 'csv:export 
            {tables? : Export some tables (tables=users,posts) }
            {--a|--all-table=0: Export All tables }
            {--f|--force : Force the operation to run when csv file already exists }
    ';

    /**
     * The console command description.
     *
     * @var string
     */
    protected $description = 'Export database tables to csv file (default dir is database/csv/<file_name>.csv )';

    /**
     * Create a new command instance.
     *
     * @return void
     */
    public function __construct()
    {
        parent::__construct();
    }

    /**
     * Execute the console command.
     *
     * @return mixed
     */
    public function handle()
    {
        ///// For all table export
        if ($this->option('all-table') and !$this->argument('tables')) {
            $this->info('Exporting all tables......');
            foreach ($this->getTables() as $table) {
                $this->writeCsv(DB::table($table)->get(), $table);
            }
            $this->info('');
            $this->info('Csv files exported Successfully.');
            return true;
        }

        $tables_str = str_replace("tables=", "", $this->argument('tables'));
        $tables_arr = explode(',', $tables_str);
        $this->info('Exporting ' . join(', ', $tables_arr) . ' tables.......');

        foreach (collect($tables_arr) as $table) {
            if (Schema::hasTable($table)) {
                $this->writeCsv(DB::table($table)->get(), $table);
            } else {
                $this->warn("table $table is not exists.");
            }
        }
        return true;
    }

    protected function getTables()
    {
        $tables = [];
        foreach (DB::connection()->getDoctrineSchemaManager()->listTableNames() as $table) {
            if (Schema::hasTable($table)) {
                $tables[] = $table;
            }
        }
        return $tables;
    }

    protected function writeCsv($data, $table_name, $directory = 'database/csv')
    {
        $csv_directory = $this->checkOrMakeDir(base_path() . '/' . $directory);
        $csv_file_path = $csv_directory . '/' . $table_name . '.csv';

        ///// METHOD ONE
        /*if (!$this->hasData($data)) {
            return 1;
        }*/

        $this->isForcefully($csv_file_path);

        $csv_file_handler = fopen($csv_file_path, 'w');
        $file_headers = null;

        $bar = $this->output->createProgressBar(count($data));
        $bar->start();

        if (!$file_headers && empty($file_headers)) {
            $file_headers = $this->writeCsvHeading($csv_file_handler, $table_name);
        }

        foreach ($data as $key => $row) {
            $row = collect($row)->toArray();
            fputcsv($csv_file_handler, $row);
            $bar->advance();
        }

        fclose($csv_file_handler);
        $bar->finish();

        $this->info(" <fg=black;bg=blue>$table_name.csv</>");
        return 1;
    }

    protected function checkOrMakeDir($path)
    {
        if (!File::exists($path)) {
            if (File::makeDirectory($path)) {
                return $path;
            }
        }
        return $path;
    }

    protected function isForcefully($path)
    {
        if (
            !$this->option('force')
            && File::exists($path)
            && !$this->confirm("The $path already exists. do you want to overwrite data ?")
        ) {
            $this->line('operation canceled.');
            die();
        }
        return 1;
    }

    protected function writeCsvHeading($file_handler, $table)
    {
        /// Doctrine DBAL Enum issue resolver
        /// Unknown database type enum, MySQL57Platform may not support it #3161
        $conn = DB::connection()->getDoctrineSchemaManager();
        $conn->getDatabasePlatform()->registerDoctrineTypeMapping('enum', 'string');
        /// Ref:1 https://github.com/doctrine/dbal/issues/3161
        /// Ref:2 https://github.com/laravel/framework/issues/1186#issuecomment-17541553
        
        $column_names = array_keys($conn->listTableColumns($table));
        return fputcsv($file_handler, $column_names);
    }

    protected function hasData($data)
    {
        return (is_countable($data) and count($data) > 0);
    }
}
